manifests/pkcs11_eventmgr.pp

```diff
@@ -32,46 +32,63 @@
 #
 #   Default: {}
 #
+# @param event_opts_base
+#   Hash
+#
+#   Default: $pam_pkcs11::params::pkcs11_event_opts
+#
 # @param autostart_method
 #   String
 #
 #   Default: $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
 #
 class pam_pkcs11::pkcs11_eventmgr (
-  Boolean    $debug            = false,
-  Boolean    $daemonize        = true,
-  Integer[0] $polling_time     = 1,
-  Integer[0] $expire_time      = 0,
-  String     $pkcs11_module    = 'default',
-  Hash       $event_opts       = {},
-  String     $autostart_method = $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
+  Boolean                 $debug            = false,
+  Boolean                 $daemonize        = true,
+  Integer[0]              $polling_time     = 1,
+  Integer[0]              $expire_time      = 0,
+  Variant[Enum['default'],
+  Stdlib::AbsolutePath] $pkcs11_module      = 'default',
+  Struct[
+    card_insert => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+    ]],
+    card_remove => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+    ]],
+    expire_time => Optional[Struct[
+        on_error => Enum['ignore', 'return', 'quit'],
+        action => Array[String],
+  ]]]                     $event_opts       = {},
+  Struct[
+    card_insert => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+    ],
+    card_remove => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+    ],
+    expire_time => Struct[
+      on_error => Enum['ignore', 'return', 'quit'],
+      action => Array[String],
+  ]]                      $event_opts_base  = $pam_pkcs11::params::pkcs11_event_opts,
+  Enum['systemd_service',
+    'xdg_autostart',
+  'none']                 $autostart_method = $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
```

Comment:
How about this?
```suggestion
  Enum[
    'systemd_service',
    'xdg_autostart',
    'none',
  ]                       $autostart_method = $pam_pkcs11::params::pkcs11_eventmgr_autostart_method,
```

Reply:
Puppet-lint will complain about it. Not sure how to improve puppet-lint.

Reply:
Also the `,` on the `'none',` line is wrong puppet syntax, however if `'none'` without `,` is in its own line, puppet-lint will complain and say that `,` is missing... So `puppet-lint` doesn't really understands the puppet language it seems. So to trick `puppet-lint` the `]` needs to be in the same line as the last entry and if that is done `puppet-lint` demands to have just 2 spaces before it.

Reply:
According to the [documentation](https://puppet.com/docs/puppet/7/lang_data_array.html),

> [an] optional trailing comma is allowed between the final value and the closing square bracket.

Reply:
Yes, Arrays, however this is a `Enum` type specification and would suspect that puppet is there a bit strange as well. These kind of irregularities are common with custom DSLs...

Reply:
Ah, of course. Sorry about that.